### DISCUSSION TITLE
Exporting to layered project formats (XCF, PSD etc.)

### DISCUSSION THREAD

**Discussion by Munkus23** [VoxelCubes/PanelCleaner#79] [Ideas]
it would be nice to export the cleaned to psd or other image editing software with the mask etc already layered so that the editing fixing and typesetting is all in one

**Comment by VoxelCubes** [comment]
I'll investigate the feasibility of supporting gimp's xcf and photoshop's psd files. I researched supporting clip studio's format but it's so absurdly proprietary that I can't support that one.

**Reply by Munkus23** [reply]
I see i hope some progress made but keep it up i appreciate your work as a one man scanlator thank you

**Reply by VoxelCubes** [reply]
I tried out all the gimp and psd python tools, even looking into the imagemagick tool, but unfortunately the support for both formats is lacking. There are lots of tools for reading these formats, but writing them, well, there is not much. 

Exporting to gimp's xcf looks to be in the works with [gimpformats](https://pypi.org/project/gimpformats/) but doesn't work yet. Maybe in the future. 

As for Photoshop's psd files, they are extremely hard to write by external tools, due to adobe not really accommodating for that. They only want photoshop to be the thing that writes psd files. So no tools for that exist, like, at all. That will never happen, unfortunately. Same situation as clip studio paint's format, sad.

Good luck scanlating, it's tough but important work!

**Reply by Munkus23** [reply]
Oh that's too bad good work though and thanks for looking this through

**Reply by Munkus23** [reply]
I've found documentation about psd tools here is the link if you haven't seen this https://psd-tools.readthedocs.io/en/latest/usage.html

Btw balloon translator has a psd export script 
![Screenshot_2024-04-17-04-11-36-629_com android chrome](https://github.com/VoxelCubes/PanelCleaner/assets/103383636/0b8ae226-5dc0-47df-87bb-d55f3929a381)

**Reply by VoxelCubes** [reply]
I appreciate the enthusiasm, good work! Unfortunately, I already looked at psd tools back then and found out that it can only be used to read psd files, not really modify them. In particular, it can't add layers, which would be the whole point here.

As for the ballons thing, that's a photoshop script. They are written in Javascript (a language I can't use) and require photoshop itself to execute (a program I can't use), hence why it is able to work with the file format.
Maybe someone else with the required knowledge could make a similar script though.

For me at least a dead end. I'll reopen the discussion so someone can find this though, might get lucky.

**Comment by Munkus23** [comment]
is this script applicable?

from PIL import Image
from psd_tools import PSDImage, Layer

def images_to_psd(image_files, output_psd_file):
    # Create an empty base image (RGBA)
    base_image = Image.new('RGBA', (1000, 1000), (255, 255, 255, 0))  # Adjust size as needed

    # Initialize layers list
    layers = []

    for image_file in image_files:
        # Open each image
        img = Image.open(image_file).convert('RGBA')

        # Create a new layer
        layer = Layer.from_image(img)
        layers.append(layer)

    # Create a PSD image with the layers
    psd = PSDImage.new(base_image)
    for layer in layers:
        psd.append(layer)

    # Save as PSD
    psd.save(output_psd_file)

# Example usage
image_files = ['image1.png', 'image2.png', 'image3.png']  # Replace with your image file paths
output_psd_file = 'output.psd'
images_to_psd(image_files, output_psd_file)

**Reply by VoxelCubes** [reply]
See #110 for the pending implementation. It needs some extra work to be upstreamed to psd-tools itself for correct layer handeling, so it's currently blocked by that. But it'll come soon, thankfully.

**Reply by Munkus23** [reply]
thank god

**Reply by VoxelCubes** [reply]
All thanks goes to [@al3ks1s](https://github.com/al3ks1s) ; )

**Comment by VoxelCubes** [comment]
@Munkus23 version 2.9 is out now with psd support! They even work perfectly fine when opened with gimp. 

Alexis, who implemented the psd support, hopes to investigate clip studio paint support some time in the future, and if gimp export has a working package, I can add that too sometime.

**Comment by Munkus23** [comment]
ooohhhh thanks alot good work i hope you have an amazing day thanks once again now i will pump out chaps
